src/Microsoft.VisualStudio.Services.Agent/HostContext.cs

```diff
@@ -169,9 +176,7 @@ public virtual string GetDirectory(WellKnownDirectory directory)
                     break;
 
                 case WellKnownDirectory.Diag:
-                    path = Path.Combine(
```

Comment:
This contradicts current behavior of GetDirectory (it returns absolute path) - could it make sense to make it conditional here instead (as for example, it is made currently for in [line](https://github.com/microsoft/azure-pipelines-agent/blob/master/src/Microsoft.VisualStudio.Services.Agent/HostContext.cs#L231))?

Reply:
At this point it could be either the AGENT_DIAGLOGPATH or the WORKER_DIAGLOGPATH or whatever was passed as hostType when the HostContext was initialized, not sure if AgentKnobs offer any such flexibility to get the setting, so I do not think that way it would work.

Reply:
@anatolybolshakov

Reply:
@teapothat this creates a duplication which makes it more error-prone in future ('_DIAGLOGPATH' is presented in knob and here), knobs mechanism is recommended way for such cases at the moment.
There are only 2 host types at the moment (although it is used as enumeration and will probably need to some refactoring here). You could probably also add some other method to get logs path depending on host type and use it here.

Reply:
@teapothat no, you can just probably create another agent knob for worker logs location.

> There is also an L1Test host type used in tests.

Yes, see it also - so 3 host types now 😃 For 'L1Test' - this is for tests only, so I don't think that it make sense to have such ability for this

Reply:
@anatolybolshakov so I looked into creating a new KnobSource that could potentially take the hostType form context (it would have to be a specific type of context) but it looks like it would not help as much as in the DiagnosticLogManager the whole point is that we do not have access to the agent context?  

Other options would be to declare a separate WorkerDiagLogKnob and an L0TestDiagLogKnob and just use all of these in a big if statement in the GetDirectory depending on the hostType (which would have to be saved as parameter). 

These are the options I can think of, but let me know if you have other suggestions?

Reply:
@teapothat yes, separate WorkerDiagLogKnob looks good - I would also introduce separate method which returns log path and taking hostType as parameter (it will probably just select it via if or switch statement using knob values)
I believe there is no need to create L0TestDiagLogKnob - not sure if there's the case to set it up for L0 tests

Reply:
@anatolybolshakov I've added the changes as discussed, would you be so kind to review them one more time?